Move to js-macaroon library

jacohend opened the pull request:
For cross-language compatibility

bucko13 committed:
add test for lsat fixture and change challenge parsing

bucko13 committed:
cleanup

Commit:
convert to go-macaroon/js-macaroon library

Commit:
fixed tests

Commit:
all tests passing

Commit:
add dist temporarily for testing

bucko13 commented on .eslintrc:
```diff
@@ -14,6 +14,8 @@
     "class-methods-use-this": 0,
     "import/no-named-as-default": 0,
     "require-atomic-updates": "warn",
+    "no-unused-vars": ["warn", { "varsIgnorePattern": "^_" }],
```
Were both needed? the typescript one didn't pick it up on its own?

jacohend replied:
It didn't compile until I included both! Not sure why.

bucko13 commented on dist/caveat.d.ts:
```diff
@@ -0,0 +1,71 @@
+import { CaveatOptions, Satisfier } from './types';
```
The dist directory shouldn't need to be checked in as this is only for publishing the post-processing code. the `prepublishOnly` npm script handles all of those steps. Keeping it out of source control just helps to keep the diffs cleaner.

jacohend replied:
I found that I couldn't use the library by including the github link in another project's deps unless I included the dist directory. Since merging this in and publishing it to npm was premature, I had to add the dist file. I can remove it as soon as this is approved.

bucko13 replied:
Good catch. Looks like the npm scripts need to be a little more... forgiving? robust? I think a `prepare` script should avoid this problem in the future [reference](https://blog.jim-nielsen.com/2018/installing-and-building-an-npm-package-from-github/). I'll add that in a separate PR that removes these as well

bucko13 commented on package.json:
```diff
@@ -21,11 +21,12 @@
   },
   "dependencies": {
     "@types/node": "^13.1.0",
+    "@types/sjcl": "^1.0.29",
```
what was this for?

jacohend replied:
hmm, looks like it's there for bitarrays in the macaroons file. I didn't add it. I wonder if an IDE did?

jacohend replied:
Removing it seemed to work. May have been a dep removed during development.

bucko13 commented on src/caveat.ts:
```diff
@@ -3,13 +3,23 @@
+
+const utf8Encoder = new TextEncoder();
+const isValue = (x: string | null | undefined) => x !== undefined && x !== null;
+const stringToBytes = (s: string | null | undefined) => isValue(s) ? utf8Encoder.encode(s) : s;
```
Might be nice to just put these in `helpers.ts` especially as there are a couple of other encoding/text related helpers in there.

bucko13 commented on src/caveat.ts:
```diff
@@ -232,37 +240,30 @@ export function verifyFirstPartyMacaroon(
+    if (satisfiers) {
+      if (!Array.isArray(satisfiers)) satisfiers = [satisfiers]
+      for (const satisfier of satisfiers) {
+        if (satisfier.condition !== caveat.condition) return "not satisifed"
```
is returning "not satisfied" rather than false something required for compatibility? This feels like more of a golang-y pattern to me (returning error string which needs to be checked for error checking and validation) and also could break expectations with any downstream dependents if they are expecting a falsey value for determining if a condition has been satisfied or not, i.e. an unsatisfied caveat could be interpreted as true since the string isn't falsey.

bucko13 replied:
Nevermind, I'm seeing below that this is just the function that's now passed to the new libraries verify function, so I'm assuming this is just what its expectations are. Leaving these comments for posterity/context and to help for my own confirmation, but I think this looks fine! Sorry :D

bucko13 commented on src/lsat.ts:
```diff
@@ -147,12 +152,16 @@ export class Lsat extends bufio.Struct {
+      return 0
+    }
+    for (const cav of caveatPackets) {
+      if (cav.i == undefined) {
```
So the caveat data is stored in an `i` property? Is there any significance to that notation do you think? It feels so arbitrary

jacohend replied:
Correct. I'm not sure why. I'm wondering if it's autogenerated object code.

bucko13 commented on tests/lsat.spec.ts:
```diff
@@ -110,6 +125,11 @@ describe('LSAT Token', () => {
+        challenge:
+          'macaroon="AgESMy4xMzYuMTc4LjE1OjM0MjM4AkIAAD2b0rX78LATiVo8bKgHuurefeF5OeX2H5ZuacBIK3+RAR1PKU1oZpfCZFib4zdDoj0pOpgPmhtuzNllU+y//D0AAAYgcWFs9FIteCzpCcEPSwmXKBpcx97hyL5Yt99cbLjRHzU=", invoice="lntb20n1psza5dwpp58kda9d0m7zcp8z2683k2spa6at08mcte88jlv8ukde5uqjpt07gsdzjfp85gnpqd9h8vmmfvdjjqurp09kk2mn5ypn8ymmdyppksctfdecx76twwssyxmmjv5sxcmny8gcnqvps8ycqzpgsp5m7xru8dlhrhmwjp8gynsj2l9mwan2jk52ah ...
+        name: 'weird padding',
+      },*/
```
cleanup?

bucko13 commented on tests/lsat.spec.ts:
```diff
@@ -187,11 +207,15 @@ describe('LSAT Token', () => {
+    const mac = Macaroon.importMacaroon(rawMac)
+    const originalcavs = originalMac._exportAsJSONObjectV2().c
+    const cavs = mac._exportAsJSONObjectV2().c
+    if (cavs == undefined || originalcavs == undefined) {
```
Why skip in a test? I'm kind of skimming so I'm probably missing something, but it feels strange to just skip. If the caveat is missing doesn't that mean there's something wrong with the test?

jacohend replied:
The tests work without it- the typescript IDE linting seems to want the check.

bucko13 commented on tests/utilities.ts:
```diff
@@ -1,22 +1,22 @@
 
-export class BuilderInterface extends MacaroonsBuilder {}
-export function getTestBuilder(secret: string): BuilderInterface {
+export function getTestBuilder(secret: string) {
```
Does the typescript not require a return type?

jacohend replied:
Type inference is used when explicit types aren't in place. I also had some issues trying to explicitly list the new macaroon class as a return type, so when I found out that this worked well, I left out the return type.

bucko13 reviewed (changes requested):
Looks great! Just a few comments and questions. 

Not really sure why the CI build is failing, but this might be a hint: https://travis-ci.community/t/run-yarn-with-frozen-lockfile-when-yarn-lock-is-present/3396
So, possibly need some change to the lock file?

Reading through the code, it also occurs to me that it might be helpful to just have a bit of documentation in the README of the underlying macaroon library being used, in part just to help future us in the event that anything has to change. I think indicating the library being used and examples of the primary API interactions (e.g. decoding/encoding macaroons and caveats). 

Curious also if this was tested with boltwall yet?

jacohend committed:
address PR comments; move helpers to helper file, remove unnecessary checks

jacohend committed:
update lock file

jacohend committed:
add back in sjcl

jacohend commented:
@bucko13 I've addressed most of your comments. I'm also attempting to fix the lockfile issue. 

I have tested this with the go implementation of the boltwall flow on the Core server side, and in Gateway on the client side. I immediately saw working cross-language behavior where previously it seemed that the macaroon library was only compatible with itself. 

I'll see about adding a note about the macaroon lib choice, as well as a link to some examples.

bucko13 commented:
Good stuff! And thanks for testing with everything that's dependent on it. I'll merge in and make the change for using as a github dependency. Feel free to make the readme updates whenever you're ready.
